### Antibody escape prediction of mutated SARS-CoV-2 variants.
Authors:

Alexander Gavrilenko
Daria Balashova

**Problem statement:**

Since the origin of SARS-Cov2, there has been enough data on RBD variants and their effect on the immune system - whether antibodies bind this variant of RBD or not (binary classification problem).

Machine learning models may be used for predictive profiling on current and prospective variants RBD guide the development of therapeutic antibody treatments and vaccines for COVID-19.

Detailed molecular analysis has revealed that many neutralizing antibodies to SARS-CoV-2 share sequence and structural features [Joseph M. Taft](https://www.sciencedirect.com/science/article/pii/S0092867422011199?via%3Dihub). It can be used as a additional data source for pretraining machine learning models for prediction binding RBD with limited labeled data of the antibody of interest. 

In this part of project I implemented fine-tuning approach adding a feature to the RBD sequence with which antibody it interacts. Fine-tuning model used for prediction of binding by therapeutic antibody REGN33.

**Usage**

Create, activate a virtual environment and install needed libraries.

```bash
python3.10 -m venv <virtual-environment-name>
source env/bin/activate
pip install -r requirements.txt
```

**Additional data** 

Load datasets of RBD sequences with physical features.

1. In a root directory:

 ```bash
mv dataset
 ```

2. Click to the links and download two files:
  phys_train: [https://disk.yandex.ru/d/PlJgrISLlixYzA](https://disk.yandex.ru/d/PlJgrISLlixYzA)

  phys_test: [https://disk.yandex.ru/d/hXVJLFn4iyWztw](https://disk.yandex.ru/d/hXVJLFn4iyWztw)

**Data description:**

 ‘Label’ column: labels corresponding to whether the antibody binds this variant of RBD or not

 ‘Antibody’ column: type of therapeutic antibody or ACE2 protein.

  *Datasets*
1. Already one-hot encoded RBD sequences and having physico-chemical features:

   [phys_train.csv](./dataset/phys_train.csv)
  
   [phys_test.csv](./dataset/phys_test.csv)
  
   ‘repr’ column: Every aminoacid in the sequence one-hot encoded and has twenty biochemical and physical features extracted from aaindex         database  [https://www.genome.jp/aaindex/](https://www.genome.jp/aaindex/)

2. Just RBD sequences strings:

   [whole_train.csv](./dataset/whole_train.csv)
   
   [whole_test.csv](./dataset/whole_test.csv)
   
   ‘junction_aa’ column: RBD mutation sequences interacted with four therapeutic antibodies and ACE2 protein

**Utils module**


[process_data.py](./utils/process_data.py) functions for processing data

[training_nn.py](./utils/training_nn.py) functions for training neural network models

[evaluation.py](./utils/evaluation.py) functions for testing neural network models

[models.py](./utils/models.py) classes of architectures neural network models

[dataset.py](./utils/dataset.py) classes of Pytorch datasets

**Notebooks:**

[clas_ML.ipynb](./notebooks/clas_ML.ipynb) EDA, training boosting classifiers for antibody escape prediction by REGN33.

[DL_full_df.ipynb](./notebooks/DL_full_df.ipynb) training neural networks with full data of REGN33.

[limit_data.ipynb](./notebooks/limit_data.ipynb) training neural networks with limited data of REGN33.

[CNN_tr_learn.ipynb](./notebooks/CNN_tr_learn.ipynb) fine-tuning of CNN model using limited data of REGN33.

[RNN_tr_learn.ipynb](./notebooks/RNN_tr_learn.ipynb) fine-tuning of RNN model using limited data of REGN33. 

**Results:**

An RNN model typically consists of two main components: the recurrent block and the linear block.
The recurrent block is composed of identical architecture cells that take the current token in the sequence and an information vector from previous cells as input. This allows the RNN to process each amino acid in the protein sequence sequentially. The output of the recurrent block provides a useful representation of the entire RBD mutation sequence, which is then fed into the linear block for classifying antibody escape.

![image](https://github.com/GavrilenkoA/ML_mutational_learning/assets/92908421/c49e34b6-6d58-49d5-be58-638f0b49347a)



Training RNN model consists of two stages:
1. Pre-training using background antibodies. Freeze weights of recurrent block.
2. Fine-tuning using limited data of REGN33 antibody.

Base model trained using only REGN33 limited data.

RNN model performance using combined data of four antibodies:


![image](https://github.com/GavrilenkoA/ML_mutational_learning/assets/92908421/a9dee58e-127f-4e87-afcf-511fc40ff9cf)



Then I provided 40 experiments, sampling different random 100 RBD sequences neutralized by REGN33 antibody for fine-tuning and compared performance of base model and fine-tuning model, using paired T-test.

![image](https://github.com/GavrilenkoA/ML_mutational_learning/assets/92908421/8344ce73-e615-4f85-91b9-c0924f836953)

It can be seen, the fine-tuned model demonstrates more accurate and stable predictions.

Our group tried other machine learning methods too. The results and code can be found here [repo](https://github.com/grigorievaekaterina/BI_antibodies_project/tree/main)

#### Conclusion: 
Therapeutic antibodies share similar features - they target the conserved epitopes on RBD and mutations in the same manner reflected by binding RBD.
They can be used as a data source for pre-training ML models for prediction antibody escape with limited labeled data of antibody of interest.

#### References:
Taft, J. M., Weber, C. R., Gao, B., Ehling, R. A., Han, J., Frei, L., Metcalfe, S. W., Overath, M. D., Yermanos, A., Kelton, W., & Reddy, S. T. (2022). Deep mutational learning predicts ACE2 binding and antibody escape to combinatorial mutations in the SARS-CoV-2 receptor-binding domain. Cell, 185(21), 4008–4022.e14. https://doi.org/10.1016/j.cell.2022.08.024

